proposals/003-authentication-api.md

```diff
@@ -0,0 +1,650 @@
+}
+```
+
+### APIs for Filters to produce server SASL information
+
+A SASL Initiator will be able to use the follow methods on `FilterContext` to announce a successful, or failed, authentication with a Kafka server:
+
+```java
+    /**
+     * Allows a filter
+     * to announce a successful authentication outcome with the Kafka server to other plugins.
+     * After calling this method the result of {@link #serverSaslContext()} will
+     * be non-empty for this and other filters.
+     * This method may be called multiple times over the lifetime of
+     * a session if reauthentication is required.
+     * TODO define the semantics around reauth
+     * @param saslPrincipal The authenticated principal.
+     */
+    void serverSaslAuthenticationSuccess(String mechanism, String serverName);
+
+    /**
+     * Allows a filter
+     * to announce a failed authentication outcome with the Kafka server.
+     * @param exception An exception describing the authentication failure.
+     */
+    void serverSaslAuthenticationFailure(String mechanism, String serverName, Exception exception);
+```
+
+### APIs for Filters to consume server SASL information
+
+```java
+    /**
+     * @return The SASL context for the server connection, or empty if the server
+     * has not successfully authenticated using SASL.
+     */
+    Optional<ServerSaslContext> serverSaslContext();
+```
+
+Where:
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+import java.util.Optional;
+
+import io.kroxylicious.proxy.filter.FilterContext;
+
+/**
+ * Exposes SASL authentication information to plugins, for example using {@link FilterContext#serverSaslContext()} ()}.
+ * This is implemented by the runtime for use by plugins.
+ */
+public interface ServerSaslContext {
+
+    /**
+     * The name of the SASL mechanism used.
+     * @return The name of the SASL mechanism used.
+     */
+    String mechanismName();
+
+    /**
+     * Returns the principal returned by the server.
+     * @return the principal returned by the server,
+     * or empty if the SASL mechanism used does not support mutual authentication.
+     */
+    String serverId();
+
+}
+```
+
+### APIs for using server principals in a generic way
+
+This works similarly to the client-facing equivalent:
+
+```java
+    /**
+     * Returns the authenticated principal for the server connection, or empty if the server
+     * has not successfully authenticated, or if the server authentication was not mutual.
+     * The concrete type of principal returned depends on the proxy configuration.
+     * For example,
+     * it may be a {@link javax.security.auth.x500.X500Principal} if server identity is TLS-based,
+     * or it may be a {@link SaslPrincipal} is client identity is SASL based.
+     * @return The authenticated principal for the server connection, or empty if the server
+     * has not successfully authenticated.
+     */
+    Optional<? extends Principal> serverPrincipal();
```

Comment:
The proxy is both a client and a server! I think `serverPrincipal` is really ambiguous.

`downstreamPrincipal` and `upstreamPrincipal` seem much clearer to me or even `receivedDownstreamPrincipal` and ` sentUpstreamPrincipal`.

Reply:
Agree on the ambiguity, but I think we should continue to shy away from upstream/downstream. It feels like the language and what it implies is cultural/individual and has caused confusion in the past.

Maybe the pain is that we've named the entity talking to the proxy `client`. If we have multiple things acting as clients, calling one of them 'the client' is confusing.

What if we introduced a new term for the client like `user`? And the upstream server is the `target`. And the proxy is the `proxy`. Then we'd have:

```
userClientPrincipal
proxyClientPrincipal
proxyServerPrincipal
targetServerPrincipal
```

Reply:
note, in this context I think `upstreamServer` would also be pretty unambiguous, it's just once we start having pervasive `upstream/downstream` language it gets harder to understand imo.

Reply:
hmm, maybe `upstream` and `downstream` language to describe the entities is fine in this networking context, IIRC the confusion in the past was more about how we were using it in documentation to describe the direction of flow of messages through filters.

Reply:
I've actually grown to like upstream and downstream terminology.